File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/holder/VisitorMessageViewHolder.java
package com.glia.widgets.chat.adapter.holder;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.LayerDrawable;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.glia.widgets.R;
import com.glia.widgets.UiTheme;
import com.glia.widgets.chat.model.history.VisitorMessageItem;
import com.glia.widgets.di.Dependencies;
import com.glia.widgets.helper.ResourceProvider;
import com.glia.widgets.view.configuration.chat.ChatStyle;

public class VisitorMessageViewHolder extends RecyclerView.ViewHolder {
    private final TextView content;
    private final TextView deliveredView;

    public VisitorMessageViewHolder(@NonNull View itemView, ChatStyle chatStyle, UiTheme uiTheme) {
        super(itemView);
        this.content = itemView.findViewById(R.id.content);
        this.deliveredView = itemView.findViewById(R.id.delivered_view);
        Context context = itemView.getContext();

        content.setBackgroundTintList(ColorStateList.valueOf(Color.parseColor(chatStyle.visitorMessage.layer.getBackgroundColor())));

        // Set corner radius
        GradientDrawable drawable = (GradientDrawable) ContextCompat.getDrawable(context, R.drawable.bg_message);
        if (drawable != null) {
            ResourceProvider resourceProvider = Dependencies.getResourceProvider();
            drawable.setCornerRadius(resourceProvider.convertDpToPixel(chatStyle.visitorMessage.layer.getCornerRadius()));
        }
        content.setBackground(drawable);

        content.setTextColor(Color.parseColor(chatStyle.visitorMessage.text.foregroundColor));
        content.setTextSize(chatStyle.visitorMessage.text.getTextSize());
        if (uiTheme.getFontRes() != null) {
            Typeface fontFamily = ResourcesCompat.getFont(context, uiTheme.getFontRes());
            content.setTypeface(fontFamily);
            deliveredView.setTypeface(fontFamily);
        }
        deliveredView.setTextColor(ContextCompat.getColor(context, uiTheme.getBaseNormalColor()));
    }

    public void bind(VisitorMessageItem item) {
        content.setText(item.getMessage());
        deliveredView.setVisibility(item.isShowDelivered() ? View.VISIBLE : View.GONE);

        String contentDescription = itemView.getResources().getString(
                item.isShowDelivered()
                        ? R.string.glia_chat_visitor_message_delivered_content_description
                        : R.string.glia_chat_visitor_message_content_description,
                item.getMessage());
        itemView.setContentDescription(contentDescription);
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/chat/adapter/holder/VisitorMessageViewHolder.kt
package com.glia.widgets.chat.adapter.holder

import androidx.recyclerview.widget.RecyclerView
import com.glia.widgets.R
import com.glia.widgets.chat.adapter.ChatAdapter
import com.glia.widgets.chat.model.VisitorMessageItem
import com.glia.widgets.databinding.ChatVisitorMessageLayoutBinding
import com.glia.widgets.di.Dependencies
import com.glia.widgets.helper.addClickActionAccessibilityLabel
import com.glia.widgets.helper.gliaAttrFont
import com.glia.widgets.helper.removeAccessibilityClickAction
import com.glia.widgets.helper.setLocaleContentDescription
import com.glia.widgets.locale.LocaleProvider
import com.glia.widgets.locale.StringKey
import com.glia.widgets.locale.StringKeyPair
import com.glia.widgets.view.unifiedui.applyLayerTheme
import com.glia.widgets.view.unifiedui.applyTextTheme
import com.glia.widgets.view.unifiedui.theme.UnifiedTheme
import com.glia.widgets.view.unifiedui.theme.chat.MessageBalloonTheme

internal class VisitorMessageViewHolder(
    private val binding: ChatVisitorMessageLayoutBinding,
    private val onRetryClickListener: ChatAdapter.OnRetryClickListener,
    unifiedTheme: UnifiedTheme? = Dependencies.gliaThemeManager.theme,
    private val localeProvider: LocaleProvider = Dependencies.localeProvider
) : RecyclerView.ViewHolder(binding.root) {

    private val visitorTheme: MessageBalloonTheme? by lazy { unifiedTheme?.chatTheme?.visitorMessage }

    private lateinit var id: String

    init {
        applyThemeFont()
        applyUnifiedTheme()
    }

    private fun applyUnifiedTheme() {
        binding.content.applyLayerTheme(visitorTheme?.background)
        binding.content.applyTextTheme(visitorTheme?.text)
    }

    /** The bubble's background and text colours come from the layout's `?attr/gliaVisitorMessage*`. */
    private fun applyThemeFont() {
        itemView.context.gliaAttrFont()?.also { binding.content.typeface = it }
    }

    fun bind(item: VisitorMessageItem) {
        this.id = item.id

        updateStatus(item.isError, item.message)
    }

    fun updateStatus(isError: Boolean, message: String) {
        binding.content.text = message

        val contentDescriptionRes = if (isError) {
            R.string.android_chat_visitor_message_not_delivered_accessibility
        } else {
            R.string.android_chat_visitor_message_accessibility
        }

        itemView.setLocaleContentDescription(contentDescriptionRes, StringKeyPair(StringKey.MESSAGE, binding.content.text.toString()))

        if (isError) {
            itemView.addClickActionAccessibilityLabel(localeProvider.getString(R.string.general_retry))
            binding.content.addClickActionAccessibilityLabel(localeProvider.getString(R.string.general_retry))
            binding.content.setOnClickListener { onRetryClickListener.onRetryClicked(id) }
            binding.content.setTextIsSelectable(false)
            itemView.setOnClickListener { onRetryClickListener.onRetryClicked(id) }
        } else {
            itemView.removeAccessibilityClickAction()
            binding.content.removeAccessibilityClickAction()
            binding.content.setTextIsSelectable(true)
            itemView.setOnClickListener(null)
            binding.content.setOnClickListener(null)
        }
    }

}
